File: api/src/test/java/com/ruby/rubia_server/controller/CampaignControllerIntegrationTest.java
package com.ruby.rubia_server.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.ruby.rubia_server.config.AbstractIntegrationTest;
import com.ruby.rubia_server.core.entity.*;
import com.ruby.rubia_server.core.enums.*;
import com.ruby.rubia_server.core.repository.*;
import com.ruby.rubia_server.core.util.CompanyContextUtil;
import com.ruby.rubia_server.dto.campaign.ProcessCampaignDTO;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.security.test.context.support.WithMockUser;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
import static org.hamcrest.Matchers.*;

@AutoConfigureMockMvc
@Sql(scripts = "/cleanup.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
class CampaignControllerIntegrationTest extends AbstractIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private CompanyContextUtil companyContextUtil;


    @Autowired
    private CompanyRepository companyRepository;

    @Autowired
    private CompanyGroupRepository companyGroupRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private DepartmentRepository departmentRepository;

    @Autowired
    private MessageTemplateRepository messageTemplateRepository;

    @Autowired
    private CampaignRepository campaignRepository;

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private CampaignContactRepository campaignContactRepository;

    @Autowired
    private ConversationRepository conversationRepository;

    @Autowired
    private ObjectMapper objectMapper;

    private CompanyGroup companyGroup;
    private Company company;
    private User user;
    private MessageTemplate messageTemplate;

    @BeforeEach
    void setUp() {
        // Setup company group
        companyGroup = CompanyGroup.builder()
            .name("Test Group")
            .build();
        companyGroup = companyGroupRepository.save(companyGroup);

        // Setup company
        company = Company.builder()
            .name("Test Company")
            .slug("test-company")
            .contactEmail("test@company.com")
            .contactPhone("1199999999")
            .companyGroup(companyGroup)
            .build();
        company = companyRepository.save(company);

        // Setup department first
        Department department = Department.builder()
            .name("Test Department")
            .company(company)
            .build();
        department = departmentRepository.save(department);

        // Setup user
        user = User.builder()
            .name("Test User")
            .email("test@user.com")
            .passwordHash("password")
            .role(UserRole.ADMIN)
            .company(company)
            .department(department)
            .build();
        user = userRepository.save(user);

        // Setup message template
        messageTemplate = MessageTemplate.builder()
            .name("Test Template")
            .content("Test message content")
            .company(company)
            .createdBy(user)
            .build();
        messageTemplate = messageTemplateRepository.save(messageTemplate);

        // Mock company context
        when(companyContextUtil.getCurrentCompanyId()).thenReturn(company.getId());
    }

    @Test
    @WithMockUser
    @Transactional
    void shouldProcessValidExcelFileAndCreateTwoContacts() throws Exception {
        // Arrange
        MockMultipartFile excelFile = new MockMultipartFile(
            "file",
            "planilha_correta.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_correta.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("Test Campaign")
            .description("Test Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO)
        );

        // Act & Assert
        String response = mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile)
                .file(dataFile)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andDo(result -> System.out.println("Response: " + result.getResponse().getContentAsString()))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true))
            .andExpect(jsonPath("$.campaign.name").value("Test Campaign"))
            .andExpect(jsonPath("$.statistics.created").value(2))
            .andExpect(jsonPath("$.statistics.processed").value(2))
            .andExpect(jsonPath("$.statistics.duplicates").value(0))
            .andExpect(jsonPath("$.statistics.errors").value(0))
            .andReturn()
            .getResponse()
            .getContentAsString();

        // Verify customers were created
        List<Customer> customers = customerRepository.findAll();
        assertThat(customers).hasSize(2);

        Customer maura = customers.stream()
            .filter(c -> "MAURA RODRIGUES".equals(c.getName()))
            .findFirst()
            .orElseThrow();
        assertThat(maura.getPhone()).isEqualTo("+5511999999999");
        assertThat(maura.getBirthDate().toString()).startsWith("2003-09-18");

        Customer tarze = customers.stream()
            .filter(c -> "TARZE CARVALHO".equals(c.getName()))
            .findFirst()
            .orElseThrow();
        assertThat(tarze.getPhone()).isEqualTo("+5511999999992");
        assertThat(tarze.getBirthDate().toString()).startsWith("2005-08-16");

        // Verify campaign was created
        List<Campaign> campaigns = campaignRepository.findAll();
        assertThat(campaigns).hasSize(1);
        assertThat(campaigns.get(0).getTotalContacts()).isEqualTo(2);

        // Verify campaign contacts were created
        List<CampaignContact> campaignContacts = campaignContactRepository.findAll();
        assertThat(campaignContacts).hasSize(2);

        // Verify conversations were created
        List<Conversation> conversations = conversationRepository.findAll();
        assertThat(conversations).hasSize(2);
    }

    @Test
    @WithMockUser
    @Transactional
    void shouldRejectExcelFileWithDuplicatePhones() throws Exception {
        // Arrange
        MockMultipartFile excelFile = new MockMultipartFile(
            "file",
            "planilha_incorreta.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_incorreta.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("Test Campaign Incorrect")
            .description("Test Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO)
        );

        // Act & Assert
        mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile)
                .file(dataFile)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true))
            .andExpect(jsonPath("$.statistics.duplicates").value(greaterThan(0)));

        // Verify only unique contacts were created
        List<Customer> customers = customerRepository.findAll();
        assertThat(customers).hasSizeLessThanOrEqualTo(1);
    }

    @Test
    @WithMockUser
    @Transactional
    void shouldCreateTwoCampaignsWithSameContactsWithoutError() throws Exception {
        // Arrange - First campaign
        MockMultipartFile excelFile1 = new MockMultipartFile(
            "file",
            "planilha_correta.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_correta.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO1 = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("First Campaign")
            .description("First Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile1 = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO1)
        );

        // Act - First campaign
        mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile1)
                .file(dataFile1)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true));

        // Arrange - Second campaign
        MockMultipartFile excelFile2 = new MockMultipartFile(
            "file",
            "planilha_correta.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_correta.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO2 = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("Second Campaign")
            .description("Second Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile2 = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO2)
        );

        // Act - Second campaign
        mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile2)
                .file(dataFile2)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true));

        // Assert - Verify both campaigns were created
        List<Campaign> campaigns = campaignRepository.findAll();
        assertThat(campaigns).hasSize(2);
        assertThat(campaigns.stream().map(Campaign::getName))
            .containsExactlyInAnyOrder("First Campaign", "Second Campaign");

        // Verify customers exist (should be same 2 customers)
        List<Customer> customers = customerRepository.findAll();
        assertThat(customers).hasSize(2);

        // Verify campaign contacts exist for both campaigns
        List<CampaignContact> campaignContacts = campaignContactRepository.findAll();
        assertThat(campaignContacts).hasSize(4); // 2 contacts × 2 campaigns

        // Verify conversations exist for both campaigns
        List<Conversation> conversations = conversationRepository.findAll();
        assertThat(conversations).hasSize(4); // 2 contacts × 2 campaigns
    }

    @Test
    @WithMockUser
    @Transactional
    void shouldUpdateContactAndCreateTwoCampaigns() throws Exception {
        // Arrange - First campaign with original data
        MockMultipartFile excelFile1 = new MockMultipartFile(
            "file",
            "planilha_correta.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_correta.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO1 = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("Original Campaign")
            .description("Original Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile1 = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO1)
        );

        // Act - First campaign
        mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile1)
                .file(dataFile1)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true));

        // Arrange - Second campaign with updated data
        MockMultipartFile excelFile2 = new MockMultipartFile(
            "file",
            "planilha_correta_update.xlsx",
            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            new ClassPathResource("static/planilha_correta_update.xlsx").getInputStream()
        );

        ProcessCampaignDTO processCampaignDTO2 = ProcessCampaignDTO.builder()
            .companyId(company.getId())
            .userId(user.getId())
            .name("Updated Campaign")
            .description("Updated Description")
            .startDate(LocalDate.now())
            .endDate(LocalDate.now().plusDays(30))
            .sourceSystem("TEST")
            .templateIds(List.of(messageTemplate.getId()))
            .build();

        MockMultipartFile dataFile2 = new MockMultipartFile(
            "data",
            "",
            "application/json",
            objectMapper.writeValueAsBytes(processCampaignDTO2)
        );

        // Act - Second campaign with updates
        mockMvc.perform(multipart("/api/campaigns/process")
                .file(excelFile2)
                .file(dataFile2)
                .contentType(MediaType.MULTIPART_FORM_DATA))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.success").value(true));

        // Assert - Verify MAURA RODRIGUES was updated
        List<Customer> customers = customerRepository.findAll();
        Customer maura = customers.stream()
            .filter(c -> "MAURA RODRIGUES".equals(c.getName()))
            .findFirst()
            .orElseThrow();
        
        assertThat(maura.getPhone()).isEqualTo("+5511999999999");
        assertThat(maura.getBirthDate().toString()).startsWith("2003-09-18");

        // Verify both campaigns were created
        List<Campaign> campaigns = campaignRepository.findAll();
        assertThat(campaigns).hasSize(2);
        assertThat(campaigns.stream().map(Campaign::getName))
            .containsExactlyInAnyOrder("Original Campaign", "Updated Campaign");

        // Verify campaign contacts and conversations exist for both campaigns
        List<CampaignContact> campaignContacts = campaignContactRepository.findAll();
        assertThat(campaignContacts).hasSize(4); // 2 contacts × 2 campaigns

        List<Conversation> conversations = conversationRepository.findAll();
        assertThat(conversations).hasSize(4); // 2 contacts × 2 campaigns
    }
}

File: api-dotnet/Rubia.Server.Tests/Integration/CampaignControllerIntegrationTest.cs
using FluentAssertions;
using Microsoft.AspNetCore.Http;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;
using Rubia.Server.Data;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Enums;
using System.Net.Http.Headers;
using System.Text;
using Xunit;

namespace Rubia.Server.Tests.Integration;

public class CampaignControllerIntegrationTest : BaseIntegrationTest, IAsyncLifetime
{
    private HttpClient _client;
    private CompanyGroup _companyGroup;
    private Company _company;
    private User _user;
    private MessageTemplate _messageTemplate;
    private Department _department;

    public async Task InitializeAsync()
    {
        _client = CreateClient();
        await SetupTestDataAsync();
    }

    public async Task DisposeAsync()
    {
        await CleanDatabaseAsync();
        _client?.Dispose();
    }

    private async Task SetupTestDataAsync()
    {
        using var scope = Services.CreateScope();
        var context = scope.ServiceProvider.GetRequiredService<RubiaDbContext>();

        // Setup company group
        _companyGroup = new CompanyGroup
        {
            Id = Guid.NewGuid(),
            Name = "Test Group"
        };
        context.CompanyGroups.Add(_companyGroup);

        // Setup company
        _company = new Company
        {
            Id = Guid.NewGuid(),
            Name = "Test Company",
            Slug = "test-company",
            ContactEmail = "test@company.com",
            ContactPhone = "1199999999",
            CompanyGroupId = _companyGroup.Id,
            CompanyGroup = _companyGroup
        };
        context.Companies.Add(_company);

        // Setup department
        _department = new Department
        {
            Id = Guid.NewGuid(),
            Name = "Test Department",
            CompanyId = _company.Id,
            Company = _company
        };
        context.Departments.Add(_department);

        // Setup user
        _user = new User
        {
            Id = Guid.NewGuid(),
            Name = "Test User",
            Email = "test@user.com",
            PasswordHash = "password",
            Role = UserRole.ADMIN,
            CompanyId = _company.Id,
            Company = _company,
            DepartmentId = _department.Id,
            Department = _department
        };
        context.Users.Add(_user);

        // Setup message template
        _messageTemplate = new MessageTemplate
        {
            Id = Guid.NewGuid(),
            Name = "Test Template",
            Content = "Test message content",
            CompanyId = _company.Id,
            Company = _company,
            CreatedById = _user.Id,
            CreatedBy = _user
        };
        context.MessageTemplates.Add(_messageTemplate);

        await context.SaveChangesAsync();
    }

    [Fact]
    public async Task ShouldProcessValidExcelFileAndCreateTwoContacts()
    {
        // Arrange
        var filePath = Path.Combine("Resources", "planilha_correta.xlsx");
        var fileBytes = await File.ReadAllBytesAsync(filePath);

        var processCampaignDto = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "Test Campaign",
            Description = "Test Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        using var formData = new MultipartFormDataContent();
        
        // Add Excel file
        var fileContent = new ByteArrayContent(fileBytes);
        fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
        formData.Add(fileContent, "file", "planilha_correta.xlsx");

        // Add JSON data
        var jsonContent = new StringContent(JsonConvert.SerializeObject(processCampaignDto), Encoding.UTF8, "application/json");
        formData.Add(jsonContent, "data");

        // Act
        var response = await _client.PostAsync("/api/campaigns/process", formData);

        // Assert
        response.Should().BeSuccessful();

        var responseContent = await response.Content.ReadAsStringAsync();
        var result = JsonConvert.DeserializeObject<dynamic>(responseContent);

        // Verify response structure
        ((bool)result.success).Should().BeTrue();
        ((string)result.campaign.name).Should().Be("Test Campaign");
        ((int)result.statistics.created).Should().Be(2);
        ((int)result.statistics.processed).Should().Be(2);
        ((int)result.statistics.duplicates).Should().Be(0);
        ((int)result.statistics.errors).Should().Be(0);

        // Verify customers were created
        using var scope = Services.CreateScope();
        var context = scope.ServiceProvider.GetRequiredService<RubiaDbContext>();
        
        var customers = await context.Customers.ToListAsync();
        customers.Should().HaveCount(2);

        var maura = customers.FirstOrDefault(c => c.Name == "MAURA RODRIGUES");
        maura.Should().NotBeNull();
        maura!.Phone.Should().Be("+5511999999999");
        maura.BirthDate.Should().HaveValue();
        maura.BirthDate!.Value.ToString("yyyy-MM-dd").Should().StartWith("2003-09-18");

        var tarze = customers.FirstOrDefault(c => c.Name == "TARZE CARVALHO");
        tarze.Should().NotBeNull();
        tarze!.Phone.Should().Be("+5511999999992");
        tarze.BirthDate.Should().HaveValue();
        tarze.BirthDate!.Value.ToString("yyyy-MM-dd").Should().StartWith("2005-08-16");

        // Verify campaign was created
        var campaigns = await context.Campaigns.ToListAsync();
        campaigns.Should().HaveCount(1);
        campaigns.First().TotalContacts.Should().Be(2);

        // Verify campaign contacts were created
        var campaignContacts = await context.CampaignContacts.ToListAsync();
        campaignContacts.Should().HaveCount(2);

        // Verify conversations were created
        var conversations = await context.Conversations.ToListAsync();
        conversations.Should().HaveCount(2);
    }

    [Fact]
    public async Task ShouldRejectExcelFileWithDuplicatePhones()
    {
        // Arrange
        var filePath = Path.Combine("Resources", "planilha_incorreta.xlsx");
        var fileBytes = await File.ReadAllBytesAsync(filePath);

        var processCampaignDto = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "Test Campaign Incorrect",
            Description = "Test Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        using var formData = new MultipartFormDataContent();
        
        // Add Excel file
        var fileContent = new ByteArrayContent(fileBytes);
        fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
        formData.Add(fileContent, "file", "planilha_incorreta.xlsx");

        // Add JSON data
        var jsonContent = new StringContent(JsonConvert.SerializeObject(processCampaignDto), Encoding.UTF8, "application/json");
        formData.Add(jsonContent, "data");

        // Act
        var response = await _client.PostAsync("/api/campaigns/process", formData);

        // Assert
        response.Should().BeSuccessful();

        var responseContent = await response.Content.ReadAsStringAsync();
        var result = JsonConvert.DeserializeObject<dynamic>(responseContent);

        // Verify response structure
        ((bool)result.success).Should().BeTrue();
        ((int)result.statistics.duplicates).Should().BeGreaterThan(0);

        // Verify only unique contacts were created
        using var scope = Services.CreateScope();
        var context = scope.ServiceProvider.GetRequiredService<RubiaDbContext>();
        
        var customers = await context.Customers.ToListAsync();
        customers.Should().HaveCountLessThanOrEqualTo(1);
    }

    [Fact]
    public async Task ShouldCreateTwoCampaignsWithSameContactsWithoutError()
    {
        // Arrange - First campaign
        var filePath = Path.Combine("Resources", "planilha_correta.xlsx");
        var fileBytes = await File.ReadAllBytesAsync(filePath);

        var processCampaignDto1 = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "First Campaign",
            Description = "First Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        // Act - First campaign
        using (var formData1 = new MultipartFormDataContent())
        {
            var fileContent1 = new ByteArrayContent(fileBytes);
            fileContent1.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
            formData1.Add(fileContent1, "file", "planilha_correta.xlsx");

            var jsonContent1 = new StringContent(JsonConvert.SerializeObject(processCampaignDto1), Encoding.UTF8, "application/json");
            formData1.Add(jsonContent1, "data");

            var response1 = await _client.PostAsync("/api/campaigns/process", formData1);
            response1.Should().BeSuccessful();

            var responseContent1 = await response1.Content.ReadAsStringAsync();
            var result1 = JsonConvert.DeserializeObject<dynamic>(responseContent1);
            ((bool)result1.success).Should().BeTrue();
        }

        // Arrange - Second campaign
        var processCampaignDto2 = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "Second Campaign",
            Description = "Second Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        // Act - Second campaign
        using (var formData2 = new MultipartFormDataContent())
        {
            var fileContent2 = new ByteArrayContent(fileBytes);
            fileContent2.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
            formData2.Add(fileContent2, "file", "planilha_correta.xlsx");

            var jsonContent2 = new StringContent(JsonConvert.SerializeObject(processCampaignDto2), Encoding.UTF8, "application/json");
            formData2.Add(jsonContent2, "data");

            var response2 = await _client.PostAsync("/api/campaigns/process", formData2);
            response2.Should().BeSuccessful();

            var responseContent2 = await response2.Content.ReadAsStringAsync();
            var result2 = JsonConvert.DeserializeObject<dynamic>(responseContent2);
            ((bool)result2.success).Should().BeTrue();
        }

        // Assert - Verify both campaigns were created
        using var scope = Services.CreateScope();
        var context = scope.ServiceProvider.GetRequiredService<RubiaDbContext>();

        var campaigns = await context.Campaigns.ToListAsync();
        campaigns.Should().HaveCount(2);
        campaigns.Select(c => c.Name).Should().Contain(new[] { "First Campaign", "Second Campaign" });

        // Verify customers exist (should be same 2 customers)
        var customers = await context.Customers.ToListAsync();
        customers.Should().HaveCount(2);

        // Verify campaign contacts exist for both campaigns
        var campaignContacts = await context.CampaignContacts.ToListAsync();
        campaignContacts.Should().HaveCount(4); // 2 contacts × 2 campaigns

        // Verify conversations exist for both campaigns
        var conversations = await context.Conversations.ToListAsync();
        conversations.Should().HaveCount(4); // 2 contacts × 2 campaigns
    }

    [Fact]
    public async Task ShouldUpdateContactAndCreateTwoCampaigns()
    {
        // Arrange - First campaign with original data
        var originalFilePath = Path.Combine("Resources", "planilha_correta.xlsx");
        var originalFileBytes = await File.ReadAllBytesAsync(originalFilePath);

        var processCampaignDto1 = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "Original Campaign",
            Description = "Original Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        // Act - First campaign
        using (var formData1 = new MultipartFormDataContent())
        {
            var fileContent1 = new ByteArrayContent(originalFileBytes);
            fileContent1.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
            formData1.Add(fileContent1, "file", "planilha_correta.xlsx");

            var jsonContent1 = new StringContent(JsonConvert.SerializeObject(processCampaignDto1), Encoding.UTF8, "application/json");
            formData1.Add(jsonContent1, "data");

            var response1 = await _client.PostAsync("/api/campaigns/process", formData1);
            response1.Should().BeSuccessful();

            var responseContent1 = await response1.Content.ReadAsStringAsync();
            var result1 = JsonConvert.DeserializeObject<dynamic>(responseContent1);
            ((bool)result1.success).Should().BeTrue();
        }

        // Arrange - Second campaign with updated data
        var updatedFilePath = Path.Combine("Resources", "planilha_correta_update.xlsx");
        var updatedFileBytes = await File.ReadAllBytesAsync(updatedFilePath);

        var processCampaignDto2 = new ProcessCampaignDto
        {
            CompanyId = _company.Id,
            UserId = _user.Id,
            Name = "Updated Campaign",
            Description = "Updated Description",
            StartDate = DateOnly.FromDateTime(DateTime.Now),
            EndDate = DateOnly.FromDateTime(DateTime.Now.AddDays(30)),
            SourceSystem = "TEST",
            TemplateIds = new List<Guid> { _messageTemplate.Id }
        };

        // Act - Second campaign with updates
        using (var formData2 = new MultipartFormDataContent())
        {
            var fileContent2 = new ByteArrayContent(updatedFileBytes);
            fileContent2.Headers.ContentType = MediaTypeHeaderValue.Parse("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
            formData2.Add(fileContent2, "file", "planilha_correta_update.xlsx");

            var jsonContent2 = new StringContent(JsonConvert.SerializeObject(processCampaignDto2), Encoding.UTF8, "application/json");
            formData2.Add(jsonContent2, "data");

            var response2 = await _client.PostAsync("/api/campaigns/process", formData2);
            response2.Should().BeSuccessful();

            var responseContent2 = await response2.Content.ReadAsStringAsync();
            var result2 = JsonConvert.DeserializeObject<dynamic>(responseContent2);
            ((bool)result2.success).Should().BeTrue();
        }

        // Assert - Verify MAURA RODRIGUES was updated
        using var scope = Services.CreateScope();
        var context = scope.ServiceProvider.GetRequiredService<RubiaDbContext>();

        var customers = await context.Customers.ToListAsync();
        var maura = customers.FirstOrDefault(c => c.Name == "MAURA RODRIGUES");
        maura.Should().NotBeNull();
        maura!.Phone.Should().Be("+5511999999999");
        maura.BirthDate.Should().HaveValue();
        maura.BirthDate!.Value.ToString("yyyy-MM-dd").Should().StartWith("2003-09-18");

        // Verify both campaigns were created
        var campaigns = await context.Campaigns.ToListAsync();
        campaigns.Should().HaveCount(2);
        campaigns.Select(c => c.Name).Should().Contain(new[] { "Original Campaign", "Updated Campaign" });

        // Verify campaign contacts and conversations exist for both campaigns
        var campaignContacts = await context.CampaignContacts.ToListAsync();
        campaignContacts.Should().HaveCount(4); // 2 contacts × 2 campaigns

        var conversations = await context.Conversations.ToListAsync();
        conversations.Should().HaveCount(4); // 2 contacts × 2 campaigns
    }
}